terraform/delete.sh

```diff
@@ -0,0 +1,55 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#!/bin/bash
+
+# This script deletes the GCP project provisioned by Stackdriver Sandbox
+
+#set -euo pipefail
+set -o errexit  # Exit on error
+
+log() { echo "$1" >&2;  }
+
+get_project(){
+    gcloud projects list --filter="name:stackdriver-sandbox-*" \
+                         --format="value(projectId)"
+}
+
+# ensure the working dir is the script's folder
+SCRIPT_DIR=$(realpath $(dirname "$0"))
+cd $SCRIPT_DIR
+
+PROJECT_ID=$(get_project)
+
+# find and delete the Stackdriver Sandbox GCP project
+if [[ -z $PROJECT_ID  ]]; then
+    log "error: Stackdriver Sandbox project not found"
+    exit 1
+fi
+log "attempting to delete $PROJECT_ID"
+gcloud projects delete $PROJECT_ID
```

Comment:
consider using `terraform destroy` instead

Reply:
What benefits would terraform destroy give us over just deleting the whole project? I can think of a couple downsides:
- What if someone accidentally deletes the tfstate files? 
- using terraform requires checking for/installing terraform if it's not installed
- in general, I prefer to avoid relying external dependencies when they're not needed. gcloud is more fool-proof if someone just wants to tear everything down. Less potential erros/bugs

Reply:
The project and most of its resources were created by terraform, so I think we've already burned the external dependency bridge, at least as far as terraform is concerned. We know it's installed because if it wasn't they wouldn't have been able to set up the sandbox in the first place.

I'm not sure I agree that gcloud is less error-prone. I got an error just now trying to delete a project that I definitely own... and that project was still deleted, despite gcloud telling me I didn't have permission to delete it.

One possible advantage to using terraform is that it will delete all of the underlying resources before deleting the actual project. This means that if someone restores the project during the grace period, it'll be empty. Deleting the project with gcloud means that if you restore the project all of your resources will come back with it.

Ultimately it probably doesn't matter which tool we use. Either way we'll end up with a deleted project. But, again, one of the reasons we chose terraform was its ability to clean up after itself. Maybe that's no longer important, I've been out of the loop for a while.

Reply:
- My main concern is still if the tfstate files go missing (again, cloud shell isn't always a fully stable environment). Wouldn't that impact terraform's ability to delete what it set up?
- The dependency is still there since cloudshell installations don't persist, so we'd need to treat the environment like a blank slate
- is the advantage you mentioned really even an advantage? I'd say if a customer attempts to restore a project, they'd expect the resources to come back if that's a possibility

I'm open to using terraform destroy if it really does provide any benefits, but it still seems to me that deleting the project directly is easier and cleaner

Reply:
It's true that there'll be problems if the tfstate file disappears. I haven't had any problems with losing stuff from cloud shell homedirs, but if that's really a concern then fair enough.

Installing terraform is dead simple and fast. [Here's how we do it for the inspec workshop](https://gitlab.com/binamov/infrastructure-testing/blob/master/setup.sh#L22-24).

Resource persistence is arguable either way. I was thinking of the scenario where one person deletes the sandbox project and then another person comes along and restores it without knowing what it is and now there's resources that people are potentially being billed for? But maybe that's not a real thing?

I also think using terraform makes this script wayyyyyy simpler and cleaner, but that really just comes down to how much maintenance burden you're willing to accept.

Honestly, it probably doesn't matter that much. Like I said before, using terraform to tear down the sandbox was part of the original design but if that doesn't make sense now then I'm totally fine doing it another way. I've just had so many bad experiences with gcloud that I don't trust it any more.